### PR TITLE
optimize method in GeometricPrimitive.Torus.New ,avoid GC ( list2array )

### PR DESCRIPTION
**Pull request by Arc-huangjingtong**
# PR Details

optimize method in GeometricPrimitive.Torus.New(...) it will avoid GC form List2Array

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] **I have built and run the editor to try this change out.**

### CODE REVIEW

**Comment by IXLLEGACYIXL** [comment]
please fil the formular, we have a summary foreach release, so people can click on your PR and read what it is about without reading the code changes

**Comment by Arc-huangjingtong** [comment]
In this func , has some fix
```
- if (tessellation < 3) tessellation = 3;
+ tessellation = Math.Max(3, tessellation); //more concise
```

```
- var vertices = new List<VertexPositionNormalTexture>();
- var indices = new List<int>();

//because the array size is definite , dont new list and toArray,it will be quickly
+ VertexPositionNormalTexture[] vertices = new VertexPositionNormalTexture[stride * stride];
+ int[] indices = new int[tessellation * tessellation * 6];

+ var indexerIndices   = 0;
+ var indexerVertices  = 0;

  ...
+ vertices[indexerVertices++] = ....

+ indices[indexerIndices++] = i * stride + j;
+ indices[indexerIndices++] = i * stride + nextJ;
+ indices[indexerIndices++] = nextI * stride + j;
                        
+ indices[indexerIndices++] = i * stride + nextJ;
+ indices[indexerIndices++] = nextI * stride + nextJ;
+ indices[indexerIndices++] = nextI * stride + j;
             


- return new GeometricMeshData<VertexPositionNormalTexture>(vertices.ToArray(), indices.ToArray(),toLeftHanded) { Name = "Torus" };
+ return new GeometricMeshData<VertexPositionNormalTexture>(vertices, indices, toLeftHanded) { Name = "Torus" };
```

**Comment by Arc-huangjingtong** [comment]
> please fil the formular, we have a summary foreach release, so people can click on your PR and read what it is about without reading the code changes

yes , i done 😄

**Comment by IXLLEGACYIXL** [comment]
the pr details would be about what its changing overall 

a description

**Comment by Arc-huangjingtong** [comment]
> the pr details would be about what its changing overall公关细节将是关于它的整体变化
> 
> a description 描述

Sorry ,now i get it 😿

File: sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs [hunk 1]
```diff
@@ -117,16 +117,18 @@ public static GeometricPrimitive New(GraphicsDevice device, float majorRadius =
             /// <exception cref="System.ArgumentOutOfRangeException">tessellation;tessellation parameter out of range</exception>
             public static GeometricMeshData<VertexPositionNormalTexture> New(float majorRadius = 0.5f, float minorRadius = 0.16666f, int tessellation = 32, float uScale = 1.0f, float vScale = 1.0f, bool toLeftHanded = false)
             {
-                var vertices = new List<VertexPositionNormalTexture>();
-                var indices = new List<int>();
-
-                if (tessellation < 3)
-                    tessellation = 3;
+                tessellation = Math.Max(3, tessellation);
```
Comment (Ethereal77) [inline sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs:120]: Although it's true the `Max()` is the same as the `if`, for these simple preconditions or adjustments I think the `if` reads better and reduces the cognitive load.

File: sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs [hunk 2]
```diff
@@ -117,16 +117,18 @@ public static GeometricPrimitive New(GraphicsDevice device, float majorRadius =
...
 
                 int stride = tessellation + 1;
 
+                VertexPositionNormalTexture[] vertices = new VertexPositionNormalTexture[stride * stride];
+                int[] indices = new int[tessellation * tessellation * 6];
```
Comment (Ethereal77) [inline sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs:125]: In the original version, `var` was used on the left side of the declaration to avoid duplication, as the right side already defines the type.

Why have you opted here to specify the type in both left and right parts of the assignment?

File: sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs (hunk 1 above)
Comment (Arc-huangjingtong) [reply sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs:120]: 好吧,见仁见智,我的认知中,max更符合自然语言,并且在特殊类型例如double中,有着更加科学的判断

File: sources/engine/Stride.VirtualReality/OpenXR/OpenXRExt_FB_passthrough.cs [hunk 3]
```diff
@@ -37,7 +36,7 @@ public OpenXRExt_FB_Passthrough(XR xr, Session session, Instance instance)
 
             bool TryGetProcAddress(string n, out nint fptr)
             {
-                PfnVoidFunction function;
+                PfnVoidFunction function = default;
```
Comment (Ethereal77) [inline sources/engine/Stride.VirtualReality/OpenXR/OpenXRExt_FB_passthrough.cs:39]: This appears to be unrelated to this specific PR concerning the GeometricPrimitive.Torus.New method.

It might be more appropriate to create a separate PR if you wish to suggest a change in OpenXR.

File: sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs (hunk 2 above)
Comment (Arc-huangjingtong) [reply sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs:125]: emmm,It doesn't matter.  the better importance is  original use List,but actually,the array size is definite ,we can optimize it

File: sources/engine/Stride.VirtualReality/OpenXR/OpenXRExt_FB_passthrough.cs (hunk 3 above)
Comment (Arc-huangjingtong) [reply sources/engine/Stride.VirtualReality/OpenXR/OpenXRExt_FB_passthrough.cs]: I giveUp it

File: sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs (hunk 1 above)
Comment (Ethereal77) [reply sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs:120]: But this is not a mathematical formula where you are using `Max` as part of an expression.
This is just a precondition that adjusts the tesselation value to a minimum of `3`.

Additionally, while translators are widely available for reading in different languages, it is important to note that all issues and PRs in this repository are written in English. Therefore, please ensure your responses do not require users to use a translator for comprehension. Thanks.

File: sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs (hunk 2 above)
Comment (Ethereal77) [reply sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs:125]: Yeah, I know it doesn't matter. It's a stylistic choice.

Just asking because the original used `var` and (in my opinion) removes some redundancy and reads better.

Comment (Arc-huangjingtong) [reply sources/engine/Stride.Graphics/GeometricPrimitives/GeometricPrimitive.Torus.cs]: I modified it

**Comment by Ethereal77** [comment]
Anyway, nice change. I think it would be nice to extend this same change to the other `GeometricPrimitive`s.

**Comment by Arc-huangjingtong** [comment]
> Anyway, nice change. I think it would be nice to extend this same change to the other `GeometricPrimitive`s.总之，不错的变化。我认为将相同的更改扩展到其他 `GeometricPrimitive` s 会很好。

yes , if this pr success ,i will modified others

**Comment by IXLLEGACYIXL** [comment]
why closed?

**Comment by Arc-huangjingtong** [comment]
> why closed?

because i operate worrying ， i will recommit in beijing time（utc➕8） night

**Comment by Arc-huangjingtong** [comment]
> > why closed? 为什么关闭？
> 
> because i operate worrying ， i will recommit in beijing time（utc➕8） night因为我担心，我会在北京时间（UTC➕8）晚上重新承诺

sorry I may not be able to fulfill my agreement

**Comment by Arc-huangjingtong** [comment]
I'm a little under the weather

**Comment by IXLLEGACYIXL** [comment]
no hurry :D
